torch/_inductor/ir.py

```diff
@@ -4234,6 +4234,172 @@ def create(
+        - input_qparams = [scale, zp], assume per-tensor quantize to uint8
+        - weight_qparams = [scales, zp, axis], assume per-channel quantize to sint8
+        - output_qparams = [scale, zp, dtype], assume per-tensor quantize
+        Scales/zero points should be taken as inputs
+        Axis/dtypes are constants
+        """
+        assert len(input_qparams) == 2  # scale, zero point
+        assert len(weight_qparams) == 3  # scale, zero point, axis
+        assert len(output_qparams) == 3  # scale, zero point, dtype
+
+        self.has_bias = has_bias
+
+        weight_qparams[0].realize()
+        weight_qparams[1].realize()
+
+        # output_qparams[:2]
+        inputs.extend(weight_qparams[:2])
+
+        # inputs.extend([packed_weight,])
+        constant_args.extend(
+            [
+                input_qparams[0],
+                input_qparams[1],
+                weight_qparams[2],
+                output_qparams[0],
+                output_qparams[1],
+                output_qparams[2],
+            ]
+        )
+
+        super().__init__(layout, inputs, constant_args)
+        self.dim = dim
+
+    def codegen(self, wrapper):
+        if self.has_bias:
+            # If bias is not None, bias will inputs[2]
+            args = [x.codegen_reference() for x in self.inputs]
+            # args is: [x, weight, bias, w_scale, w_zp]
+            packed_weight = args[1]
+            bias = args[2]
+            w_scale, w_zp = args[-2], args[-1]
+            const_args = []
+            const_args.extend(self.codegen_const_args())
+            # const_args is: [stride, padding, dilation, groups, x_scale, x_zp, w_axis, o_inv_scale, o_zp, o_dtype]
+            o_dtype = const_args[-1]
+            o_inv_scale, o_zp = const_args[-3], const_args[-2]
+            w_axis = const_args[-4]
+            x_scale, x_zp = const_args[-6], const_args[-5]
+            stride = const_args[0]
+            padding = const_args[1]
+            dilation = const_args[2]
+            groups = const_args[3]
+        else:
+            # If bias is None, bias will be the first element of const_args
+            args = [x.codegen_reference() for x in self.inputs]
+            # args is: [x, weight, w_scale, w_zp]
+            packed_weight = args[1]
+            w_scale, w_zp = args[-2], args[-1]
+            const_args = []
+            const_args.extend(self.codegen_const_args())
+            # const_args is: [bias, stride, padding, dilation, groups, x_scale, x_zp, w_axis, o_inv_scale, o_zp, o_dtype]
+            o_dtype = const_args[-1]
+            o_inv_scale, o_zp = const_args[-3], const_args[-2]
+            w_axis = const_args[-4]
+            x_scale, x_zp = const_args[-6], const_args[-5]
+            bias = const_args[0]
+            stride = const_args[1]
+            padding = const_args[2]
+            dilation = const_args[3]
+            groups = const_args[4]
+
+        self.kernel = "torch.ops.quantized.qconv2d_pt2e"
```

Comment:
So this is just dispatching to a handwritten kernel, rather than using compiler to do codegen? Why is it done via inductor?

Also as I said in another PR, we should really use backend specific ops and custom libs for this purpose.

We used to do torch.ops.quantized.* because we effectively had "quantized" backend in aten. But quantized model in PT2E is represented via canonical aten ops (except for chose_qparams). Thus if you take "dq -> conv -> q" and convert that pattern to a custom op "quantized_conv", then such an op should not and does not need to be in "torch.op.quantized.*". It should really be in "torch.ops.onednn" as it is backend specific. cc: @jerryzh168

Reply:
yeah agree, I think this should be put into the `torch.ops.onednn` namespace

Reply:
> So this is just dispatching to a handwritten kernel, rather than using compiler to do codegen? Why is it done via inductor?

Not all kernels in the inductor compilation path are codegen. There are both library-based fusion kernels and codegen kernels. The former is primarily for fusions with conv and gemm ops and the latter is for memory intensive operations like pointwise and reduction.

> yeah agree, I think this should be put into the `torch.ops.onednn` namespace

Sure, let's put them under onednn namespace.

Reply:
Thanks and moved them into onednn namespace.